src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java

```diff
@@ -199,46 +196,32 @@ private static int countParams(String sql) {
     /**
      * Used to keep track of an individual handle ready for un-prepare.
      */
-    final class PreparedStatementCacheItem {
-        private final PreparedStatementHandle statementHandle = new PreparedStatementHandle();
-        private volatile SQLServerParameterMetaData parameterMetadata;
-        private volatile boolean hasExecutedSpExecuteSql;
-        volatile boolean evictedFromCache;
+    final class PreparedStatementMetadata {
+        private final SQLServerParameterMetaData parameterMetadata;
+        private final CityHash128Key hashKey;
 
-        boolean hasExecutedSpExecuteSql() {
-            return hasExecutedSpExecuteSql;
-        }
-
-        void setHasExecutedSpExecuteSql(boolean hasExecutedSpExecuteSql) {
-            this.hasExecutedSpExecuteSql = hasExecutedSpExecuteSql;
-        }
-
-        boolean hasHandle() {
-            return 0 < statementHandle.getHandle();
-        }
+        PreparedStatementMetadata(CityHash128Key hashKey, SQLServerParameterMetaData parameterMetadata) {
+            this.hashKey = hashKey;
+            this.parameterMetadata = parameterMetadata;
 
-        PreparedStatementHandle getPreparedStatementHandle() {
-            return statementHandle;
-        }
-        
-        boolean hasParameterMetadata() {
-            return null != this.parameterMetadata;
+            // keep the statement handle alive for the life of the prepared statement metadata
+            borrowStatementHandle(hashKey);
         }
 
         SQLServerParameterMetaData getParameterMetadata() {
             return parameterMetadata;
         }
 
-        void setParameterMetadata(SQLServerParameterMetaData metadata) {
-            parameterMetadata = metadata;
+        void releaseStatementHandle() {
+            returnStatementHandle(hashKey);
         }
     }
 
     /** Size of the  prepared statement handle cache */
     private int statementPoolingCacheSize = 10;
 
     /** Cache of prepared statement handles */
-    private ConcurrentLinkedHashMap<Sha1HashKey, PreparedStatementCacheItem> preparedStatementCache;
+    private ConcurrentLinkedHashMap<CityHash128Key, PreparedStatementMetadata> preparedStatementMetadataCache;
```

Comment:
@TobiasSQL Do you want me to sever the coupling between the two caches and push it to my ``metadataCache2`` branch?  Or is that something you want to look at post-6.1.8?